### PR TITLE
Disambiguate structured and abstract matrix multiplication

### LINKED ISSUE #1022: Inconsistent `*` for `AbstractMatrix` with `AbstractMatrix`, and `AbstractMatrix` with `Diagonal`, etc.?
In the current `LinearAlgebra` module, [`*` for `AbstractMatrix` with `AbstractMatrix`](https://github.com/JuliaLang/julia/blob/v1.10.0-beta1/stdlib/LinearAlgebra/src/matmul.jl#L104-L107), and [`AbstractMatrix` with `Diagonal`](https://github.com/JuliaLang/julia/blob/v1.10.0-beta1/stdlib/LinearAlgebra/src/diagonal.jl#L288-L289) use different container types:

```julia
function (*)(A::AbstractMatrix, B::AbstractMatrix)
    TS = promote_op(matprod, eltype(A), eltype(B))
    mul!(similar(B, TS, (size(A, 1), size(B, 2))), A, B)
end

(*)(A::AbstractMatrix, D::Diagonal) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag))), A, D)
```

As you see, the first method uses `B` in `similar`, while the second method uses `A` in `similar`. I could assume that usually with you multiply a diagonal matrix, you want to expand the first matrix, so `A` is used in `similar`. But why is `B` used in the first method? It caused me some problems in [the code I wrote](https://github.com/MineralsCloud/CrystallographyCore.jl):

```julia
julia> using CrystallographyCore

julia> lattice = Lattice(2, 1, 5, 90, 90, 90)
3×3 Lattice{Float64}
 2.0  0.0  0.0
 0.0  1.0  0.0
 0.0  0.0  5.0

julia> b = rand(1:2, 3, 3)
3×3 Matrix{Int64}:
 1  2  2
 2  2  2
 2  1  1

julia> lattice * b
3×3 Matrix{Float64}:
  2.0  4.0  4.0
  2.0  2.0  2.0
 10.0  5.0  5.0

julia> b * lattice
3×3 Lattice{Float64}
 2.0  2.0  10.0
 4.0  2.0  10.0
 4.0  1.0  5.0
```

Where `Lattice` is a subtype of `AbstractMatrix`. As you can see, the left and right multiplications of `Lattice` with another matrix produced different types. Is there a cleverer way to guess the container type of `*`?

### LINKED ISSUE #51354: Inconsistent `*` for `AbstractMatrix` with `AbstractMatrix`, and `AbstractMatrix` with `Diagonal`, etc.?
In the current `LinearAlgebra` module, [`*` for `AbstractMatrix` with `AbstractMatrix`](https://github.com/JuliaLang/julia/blob/v1.10.0-beta1/stdlib/LinearAlgebra/src/matmul.jl#L104-L107), and [`AbstractMatrix` with `Diagonal`](https://github.com/JuliaLang/julia/blob/v1.10.0-beta1/stdlib/LinearAlgebra/src/diagonal.jl#L288-L289) use different container types:

```julia
function (*)(A::AbstractMatrix, B::AbstractMatrix)
    TS = promote_op(matprod, eltype(A), eltype(B))
    mul!(similar(B, TS, (size(A, 1), size(B, 2))), A, B)
end

(*)(A::AbstractMatrix, D::Diagonal) =
    mul!(similar(A, promote_op(*, eltype(A), eltype(D.diag))), A, D)
```

As you see, the first method uses `B` in `similar`, while the second method uses `A` in `similar`. I could assume that usually with you multiply a diagonal matrix, you want to expand the first matrix, so `A` is used in `similar`. But why is `B` used in the first method? It caused me some problems in [the code I wrote](https://github.com/MineralsCloud/CrystallographyCore.jl):

```julia
julia> using CrystallographyCore

julia> lattice = Lattice(2, 1, 5, 90, 90, 90)
3×3 Lattice{Float64}
 2.0  0.0  0.0
 0.0  1.0  0.0
 0.0  0.0  5.0

julia> b = rand(1:2, 3, 3)
3×3 Matrix{Int64}:
 1  2  2
 2  2  2
 2  1  1

julia> lattice * b
3×3 Matrix{Float64}:
  2.0  4.0  4.0
  2.0  2.0  2.0
 10.0  5.0  5.0

julia> b * lattice
3×3 Lattice{Float64}
 2.0  2.0  10.0
 4.0  2.0  10.0
 4.0  1.0  5.0
```

Where `Lattice` is a subtype of `AbstractMatrix`. As you can see, the left and right multiplications of `Lattice` with another matrix produced different types. Is there a cleverer way to guess the container type of `*`?

### PR DESCRIPTION
**Pull request by jishnub**
The various methods of the form `*(::AbstractMatrix, ::Diagonal)` for structured matrices lead to method ambiguities in packages that define `*(::MyMatrix, ::AbstractMatrix)`. However, these methods only seem to be defined here to use the first argument to construct the destination, as opposed to the fallback method that uses the second argument. Instead of specializing `*`, we might as well specialize the call to `similar` that chooses which matrix to use to construct the destination. This reduces code duplication, as well as avoids these ambiguities altogether. This also makes the destination types of `A * B` and `B * A` identical if one of the two is a `StructuredMatrix`.

After this, for example, both of these produce sparse matrices:
```julia
julia> S = sprand(4,4,0.2); B = Bidiagonal(ones(4), ones(3), :U);

julia> S * B
4×4 SparseMatrixCSC{Float64, Int64} with 2 stored entries:
  ⋅    ⋅   0.706461  0.706461
  ⋅    ⋅    ⋅         ⋅ 
  ⋅    ⋅    ⋅         ⋅ 
  ⋅    ⋅    ⋅         ⋅ 

julia> B * S
4×4 SparseMatrixCSC{Float64, Int64} with 1 stored entry:
  ⋅    ⋅   0.706461   ⋅ 
  ⋅    ⋅    ⋅         ⋅ 
  ⋅    ⋅    ⋅         ⋅ 
  ⋅    ⋅    ⋅         ⋅ 
```
The former is a `Matrix` on master.

Close JuliaLang/LinearAlgebra.jl#1022
Address https://github.com/JuliaStats/PDMats.jl/issues/176 on nightly

This PR changes the following behavior:
```julia
julia> T = UpperTriangular(Number[i+j for i in 1:4, j in 1:4]);

julia> v = rand(4);

julia> T * v |> eltype
Float64

julia> Array(T) * v |> eltype
Any
```
to
```julia
julia> T * v |> eltype
Any
```
I'm unsure if the choice of `_init_eltype` was deliberate here (there is a test that specifically checks that `_init_eltype` isn't being used for general matrices).

### CODE REVIEW

**Comment by dkarrasch** [comment]
That is nice. I believe we have similar `*` methods in SparseArrays.jl that tell it to call similar on the first argument. We should use this mechanism over there as well.

File: stdlib/LinearAlgebra/src/matmul.jl [hunk 1]
```diff
@@ -106,8 +106,11 @@ julia> [1 1; 0 1] * [1 0; 1 1]
 """
 function (*)(A::AbstractMatrix, B::AbstractMatrix)
     TS = promote_op(matprod, eltype(A), eltype(B))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/matmul.jl:108]: Why not use `_init_eltype` here? What's the test you mentioned? The issue is that the `_init_eltype` mechanism exists to handle unitful matrices, where the return eltype is not obtained from promotion.

Comment (jishnub) [reply stdlib/LinearAlgebra/src/matmul.jl:108]: This is the test:
https://github.com/JuliaLang/julia/blob/d90fa45c6ac5da8398e6864128579df48edbf00c/stdlib/LinearAlgebra/test/matmul.jl#L934-L941
If we use `_init_eltype`, the test that used to throw actually passes now, as `C` becomes a `6×0 Matrix{Float64}`. However, perhaps this test may be updated?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/matmul.jl:108]: IIRC, then `_init_eltype` falls back to `promote_op(matprod, eltype(A), eltype(B))` for eltypes other than `Number`, but for `Number`s allows to compute the unit of the product. So it should be "superior" to `promote_op`-only. And in the test, from what I understand, the issue was to make sure that the product be computed. To make sure to test code where `C` doesn't have a zero, one could modify it to

```julia
C = collect(Any, zeros(r, 0))
mul!(C, A, B)
@test size(A * B) == (r, 0)
```

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/matmul.jl:108]: I'd suggest to do this:

```suggestion
    TS = _init_eltype(matprod, eltype(A), eltype(B))
```

In fact, we should replace all occurrences of `_init_eltype(*` by `_init_eltype(matprod` and remove the specialized method for the `*`-case. Moreover, we should have tests with unitful matrices (full and triangonal). I'm surprised that CI passes given that the eltype promotion for triangular matrices is not supposed to work without ` _init_eltype`  for unitful matrices.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/matmul.jl:108]: I just double-checked. It seems that things have greatly improved after v1.9 in terms of return type inference. It seems like `promote_op` works across unitful and unitless eltypes.

**Comment by dkarrasch** [comment]
I have pushed a commit that removes the clumsy `_init*` stuff, since unitful arithmetic seems to work just fine with `promote_op(matprod, eltype(A), eltype(B))`. The inconsistency between triangular and plain matrices was unintended. The promotion mechanism was introduced in a big triangular overhaul, and back then, I didn't extrapolate to the whole LinearAlgebra library.

**Comment by dkarrasch** [comment]
Hm, in fact, `matprod_dest` could be equally well used in the solves, seems like...?

**Comment by jishnub** [comment]
Looks possible. I'll try to get to this in a couple of days. Thanks for making this simpler!

**Comment by dkarrasch** [comment]
BTW, note #35166.

**Comment by jishnub** [comment]
Is there anything left to be done here?

**Comment by dkarrasch** [comment]
I have some local stuff for the determination of the eltype of the destination array, but that could go into another PR.

**Comment by singularitti** [comment]
Nice work @jishnub @dkarrasch, thank you!

**Pull request merged**